### PR TITLE
Use context managers more often

### PR DESCRIPTION
**Pull request by QuLogic**
I realize that meson is mostly an executable that is intended to be run for a very short while, but there are a lot of opened and dangling file handles. It's generally a problem for Windows where an open file handle can prevent some other process from accessing the same file. Plus too many open files might run into some system limit (though it's pretty unlikely on most modern systems.)

I'm not sure about `generate_unity_files` because there don't seem to be any unity build tests?

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -1386,44 +1388,55 @@ def scan_fortran_module_outputs(self, target):
...
-                mod_source_file = tdeps[usename]
-                # Check if a source uses a module it exports itself.
-                # Potential bug if multiple targets have a file with
-                # the same name.
-                if mod_source_file.fname == os.path.split(src)[1]:
-                    continue
-                mod_name = compiler.module_name_to_filename(usematch.group(1))
-                mod_files.append(os.path.join(dirname, mod_name))
+        with open(src) as f:
+            for line in f:
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: `for line in f.readlines()` I suppose

Comment (QuLogic) [inline mesonbuild/backend/ninjabackend.py]: Not necessary; that's how it was before.

**Comment by jpakkane** [comment]
The files are actually closed automatically due to the way cpython does refcounting. This is still good stuff just to be more precise, but due to priorities it might take me a few days to get this reviewed. Thanks.

**Comment by QuLogic** [comment]
Rebased to get get CI refreshed.

**Comment by QuLogic** [comment]
Rebased to fix conflicts.

**Pull request merged**